bindings/java/src/main/native/jucx_common_def.cc

```diff
@@ -204,11 +221,25 @@ UCS_PROFILE_FUNC(jobject, process_request, (request, callback), void *request, j
 
     // If request is a pointer set context callback and rkey.
     if (UCS_PTR_IS_PTR(request)) {
-        if (callback != NULL) {
-            ((struct jucx_context *)request)->callback = env->NewGlobalRef(callback);
+        struct jucx_context *ctx = (struct jucx_context *)request;
+        ucs_spin_lock(&ctx->lock);
+        if (ucs_unlikely(ctx->is_completed && callback != NULL)) {
+            // Whether ucp_tag_recv_nb, when recv callback was called already.
+            // Or recv_callback was called from another progress thread earlier.
+            // But we need to call java callback here.
+            if (ctx->status == UCS_OK) {
+                call_on_success(callback, jucx_request);
+            } else {
+                call_on_error(callback, ctx->status);
+            }
+            jucx_request_reset(request);
+        } else {
+            if (callback != NULL) {
+                ctx->callback = env->NewGlobalRef(callback);
+            }
```

Comment:
if request was completed and callback is NULL, where is it released (ucp_request_free) ?